napari/_qt/_tests/test_threading.py

```diff
@@ -0,0 +1,175 @@
+
+    newfunc = threading.as_generator_function(func)
+    assert inspect.isgeneratorfunction(newfunc)
+    assert list(newfunc()) == [None]
+
+
+# qtbot is necessary for qthreading here.
+# note: pytest-cov cannot check coverage of code run in the other thread.
+def test_thread_worker(qtbot):
+    """Test basic threadworker on a function"""
+
+    def func():
+        return 1
+
+    def test(v):
+        assert v == 1
+
+    thread_func = threading.thread_worker(func, connect={'returned': test})
+    worker = thread_func()
+    assert isinstance(worker, threading.FunctionWorker)
+
+
+def test_thread_generator_worker(qtbot):
+    """Test basic threadworker on a generator"""
+
+    def func():
+        yield 1
+        yield 1
+        return 3
+
+    def test_return(v):
+        assert v == 3
+
+    def test_yield(v):
+        assert v == 1
+
+    thread_func = threading.thread_worker(
+        func, connect={'returned': test_return, 'yielded': test_yield}
+    )
+    worker = thread_func()
+    assert isinstance(worker, threading.GeneratorWorker)
+
+
+def test_thread_raises(qtbot):
+    """Test exceptions get returned to main thread"""
+
+    def func():
+        yield 1
+        yield 1
+        raise ValueError('whoops')
+
+    def handle_raise(e):
+        assert isinstance(e, ValueError)
+        assert str(e) == 'whoops'
+
+    thread_func = threading.thread_worker(
+        func, connect={'errored': handle_raise}
+    )
+    worker = thread_func()
+    assert isinstance(worker, threading.GeneratorWorker)
+
+
+def test_multiple_connections(qtbot):
+    """Test the connect dict accepts a list of functions, and type checks"""
+
+    def func():
+        return 1
+
+    def test1(v):
+        assert v == 1
+
+    def test2(v):
+        assert v == 1
+
+    thread_func = threading.thread_worker(
+        func, connect={'returned': [test1, test2]}
+    )
+    worker = thread_func()
+    assert isinstance(worker, threading.FunctionWorker)
+
+    # they must all be functions
+    with pytest.raises(TypeError):
+        threading.thread_worker(func, connect={'returned': ['test1', test2]})()
+
+    # they must all be functions
+    with pytest.raises(TypeError):
+        threading.thread_worker(func, connect=test1)()
+
+
+def test_create_worker():
+    """Test directly calling create_worker."""
+
+    def func(x, y):
+        return x + y
+
+    worker = threading.create_worker(func, 1, 2)
+    assert isinstance(worker, threading.WorkerBase)
+
+    with pytest.raises(TypeError):
+        _ = threading.create_worker(func, 1, 2, _worker_class=object)
+
+
+# note: pytest-cov cannot check coverage of code run in the other thread.
+# this is just for the sake of coverage
+def test_thread_worker_in_main_thread():
+    """Test basic threadworker on a function"""
+
+    def func(x):
+        return x
+
+    thread_func = threading.thread_worker(func)
+    worker = thread_func(2)
+    assert worker.work() == 2
```

Comment:
What exactly in this code is causing it to be run on the main thread?

Reply:
the fact that we call `worker.work()` instead of `worker.start()`... will add that to the docs

Reply:
Lovely!